File: app/src/main/java/com/ytempest/wanandroid/activity/main/project/ProjectFrag.java
package com.ytempest.wanandroid.activity.main.project;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.material.tabs.TabLayout;
import androidx.viewpager.widget.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import com.ytempest.layoutinjector.annotation.InjectLayout;
import com.ytempest.tool.util.DataUtils;
import com.ytempest.wanandroid.R;
import com.ytempest.wanandroid.base.fragment.LoaderFrag;
import com.ytempest.wanandroid.base.load.ViewType;
import com.ytempest.wanandroid.http.bean.ProjectClassifyBean;

import java.util.List;

import butterknife.BindView;

/**
 * @author heqidu
 * @since 2020/6/23
 */
@InjectLayout(R.layout.frag_project)
public class ProjectFrag extends LoaderFrag<ProjectPresenter> implements IProjectContract.View {

    private static final String TAG = ProjectFrag.class.getSimpleName();

    @BindView(R.id.group_project_tab)
    TabLayout mTabLayout;
    @BindView(R.id.vp_project_content)
    ViewPager mViewPager;
    private ProjectClassifyAdapter mAdapter;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mViewPager.setOffscreenPageLimit(3);
        mAdapter = new ProjectClassifyAdapter(getActivity().getSupportFragmentManager());
        mViewPager.setAdapter(mAdapter);

        mTabLayout.setupWithViewPager(mViewPager);
        mTabLayout.setTabMode(TabLayout.MODE_SCROLLABLE);
        mTabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {

            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                setSelectedTab(tab, true);
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
                setSelectedTab(tab, false);
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
            }
        });

        getLoader().showView(ViewType.LOAD);
        mPresenter.getProjectClassify();
    }

    @Override
    protected void onReloadClick() {
        super.onReloadClick();
        mPresenter.getProjectClassify();
    }

    @Override
    public void onProjectClassifyReceived(List<ProjectClassifyBean> list) {
        getLoader().hideAll();
        mAdapter.display(list);
        updateTabView(list);
    }

    @Override
    public void onProjectClassifyFail(int code) {
        getLoader().showView(ViewType.ERR);
    }

    private void updateTabView(List<ProjectClassifyBean> list) {
        LayoutInflater inflater = LayoutInflater.from(getContext());
        for (int i = 0, count = DataUtils.getSize(list); i < count; i++) {
            TabLayout.Tab tab = mTabLayout.getTabAt(i);
            if (tab != null) {
                View view = inflater.inflate(R.layout.item_tab_classify_content, null, false);
                TextView textView = view.findViewById(R.id.tv_tab_classify_content);
                textView.setText(list.get(i).getName());
                tab.setCustomView(view);
            }
        }

        TabLayout.Tab firstTab = mTabLayout.getTabAt(0);
        if (firstTab != null) {
            setSelectedTab(firstTab, true);
        }
    }

    private void setSelectedTab(TabLayout.Tab tab, boolean selected) {
        View view = tab.getCustomView();
        if (view != null) {
            view.findViewById(R.id.tv_tab_classify_content).setSelected(selected);
        }
    }
}

File: app/src/main/java/com/ytempest/wanandroid/activity/main/project/ProjectFrag.kt
package com.ytempest.wanandroid.activity.main.project

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.google.android.material.tabs.TabLayout
import com.ytempest.wanandroid.R
import com.ytempest.framework.base.createViewModel
import com.ytempest.framework.base.fragment.MVVMFragment
import com.ytempest.wanandroid.base.load.Loader
import com.ytempest.wanandroid.base.load.ViewType
import com.ytempest.wanandroid.base.vm.EntityObserver
import com.ytempest.wanandroid.databinding.FragProjectBinding
import com.ytempest.wanandroid.http.bean.ProjectClassifyBean

/**
 * @author heqidu
 * @since 21-2-10
 */
class ProjectFrag : MVVMFragment<FragProjectBinding>(), IProjectView {

    override val viewModel by lazy { createViewModel<ProjectViewModel>() }
    private lateinit var mAdapter: ProjectClassifyAdapter

    private val loader: Loader by lazy {
        Loader(binding.root as ViewGroup).also {
            it.setReloadCall {
                it.showView(ViewType.LOAD)
                viewModel.loadProjectClassify()
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        activity?.let {
            mAdapter = ProjectClassifyAdapter(activity!!.supportFragmentManager)
            binding.pageView.offscreenPageLimit = 3
            binding.pageView.adapter = mAdapter

            binding.tabView.setupWithViewPager(binding.pageView)
            binding.tabView.tabMode = TabLayout.MODE_SCROLLABLE
            binding.tabView.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
                override fun onTabSelected(tab: TabLayout.Tab?) = setSelectedTab(tab, true)

                override fun onTabReselected(tab: TabLayout.Tab?) = setSelectedTab(tab, false)

                override fun onTabUnselected(tab: TabLayout.Tab?) = Unit
            })
        }
        initData()
    }

    private fun initData() {
        viewModel.projectClassifyResult.observe(this, EntityObserver(
            onSuccess = { entity ->
                onProjectClassifyReceived(entity.data)
            },
            onFail = { entity ->
                onProjectClassifyFail(entity.code)
            }
        ))

        loader.showView(ViewType.LOAD)
        viewModel.loadProjectClassify()
    }

    private fun setSelectedTab(tab: TabLayout.Tab?, selected: Boolean) {
        tab?.customView?.let {
            it.findViewById<View>(R.id.tv_tab_classify_content).isSelected = selected
        }
    }

    override fun onProjectClassifyReceived(list: List<ProjectClassifyBean>) {
        loader.hideAll()
        mAdapter.display(list)
        updateTabView(list)
    }

    private fun updateTabView(list: List<ProjectClassifyBean>) {
        val inflater = LayoutInflater.from(context)
        for ((idx, bean) in list.withIndex()) {
            val tab = binding.tabView.getTabAt(idx)
            val view = inflater.inflate(R.layout.item_tab_classify_content, null, false)
            view.findViewById<TextView>(R.id.tv_tab_classify_content).text = bean.name
            tab?.customView = view
        }
        val firstTab = binding.tabView.getTabAt(0)
        setSelectedTab(firstTab, true)
    }

    override fun onProjectClassifyFail(code: Int) {
        loader.showView(ViewType.ERR)
    }

}